### PR TITLE
Api

### PR DESCRIPTION
**Pull request by MikeHitrov**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Movies API # 
+A simple API, which stores information about movies.
```
Comment (svilenvelikov) [inline README.md]: Should read like: "...which provides CRUD operations over movies model" or something like this

File: data/movies.json [hunk 2]
```diff
@@ -0,0 +1 @@
+[{"id":1,"title":"Fast & Furious 4","director":"Justin Lin","releaseDate":2009,"genre":"action","actors":["Vin Diesel","Paul Walker","Michelle Rodriguez","Jordana Brewster","John Ortiz"],"createdAt":"Mon Nov 23 2020 14:21:17 GMT+0200 (CEST)","updatedAt":"Mon Nov 23 2020 14:21:17 GMT+0200 (CEST)"},{"id":2,"title":"Star Wars","director":"George Lucas","releaseDate":1977,"genre":"science fiction","actors":["Mark Hamill","Harrison Ford","Carrie Fisher","Peter Cushing","Alec Guinness"],"createdAt":"Mon Nov 23 2020 14:32:03 GMT+0200 (CEST)","updatedAt":"Mon Nov 23 2020 14:32:03 GMT+0200 (CEST)"}]
```
Comment (svilenvelikov) [inline data/movies.json]: eventualy format the file

File: models/movie.model.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
+const filename =
+  process.env.HOME + "/Desktop/Programming/Ontotext/MoviesAPI/data/movies.json";
```
Comment (svilenvelikov) [inline models/movie.model.js]: Do I have such file on my machine? Shouldn't you be using the json file which is in data?

File: helpers/helper.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+const newDate = () => new Date().toString();
+
+/**
+ * This function checks if a movie is in the object based on its id
+ *
+ * @param  {Array} array The movies array
+ * @param  {number} id The movie id
+ * @returns {object} The movie
+ */
+function mustBeInArray(array, id) {
```
Comment (svilenvelikov) [inline helpers/helper.js]: The name implies that the method checks simething and by convention such methods must return a boolean value.

File: helpers/helper.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ * @returns {string} The generated date
+ */
+const newDate = () => new Date().toString();
+
+/**
+ * This function checks if a movie is in the object based on its id
+ *
+ * @param  {Array} array The movies array
+ * @param  {number} id The movie id
+ * @returns {object} The movie
```
Comment (svilenvelikov) [inline helpers/helper.js]: This is not true. The method returns movie model if it's found in the array, but it does return an error otherwise.

File: models/movie.model.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  });
+}
+
+/**
+ * This function returns a movie by given id
+ *
+ * @param {number} id The movie id
+ * @returns {Promise} Returns the movie if present
+ */
+function getMovie(id) {
```
Comment (svilenvelikov) [inline models/movie.model.js]: I'm little confused here. The helper.mustBeInArray returns a Promise, right? Why do you need to wrap it in another promise?! As I see it, you just need to do
```javascript
function getMovie(id) {
  return helper.mustBeInArray(movies, id);
}
```

File: helpers/helper.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+/**
+ * This function writes on the JSON file
+ *
+ * @param  {string} filename The path to the file
+ * @param  {string} content The content that needs to be written
+ */
+function writeJSONFile(filename, content) {
+  fs.writeFileSync(filename, JSON.stringify(content), "utf8", (err) => {
```
Comment (svilenvelikov) [inline helpers/helper.js]: Why you need to write the file synchronously? What will be the implications if you write it asynchronously? What about if the file is big? If multiple users try to update the file, what then?

File: helpers/helper.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    if (err) {
+      console.log(err);
```
Comment (svilenvelikov) [inline helpers/helper.js]: You are effectively swallowing the error here. The API user won't know if there was and error, right?

File: models/movie.model.js (hunk 3 above)
Comment (MikeHitrov) [reply models/movie.model.js]: This is the file that is in data, but in order to write to it it is required to specify the full path.

File: helpers/helper.js (hunk 8 above)
Comment (MikeHitrov) [reply helpers/helper.js]: Is it a good idea if I write it as a record in the file, and if it is an error it will be displayed in the frontend. Or is it better to write the errors in a separate file?

File: models/movie.model.js (hunk 3 above)
Comment (svilenvelikov) [reply models/movie.model.js]: You can get the current working directory and avoid hardcodding it. Search for CWD

File: models/movie.model.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+/**
+ * This function updares a movie by given id and rewrites the JSON file
+ *
+ * @param {number} id The movie id
+ * @param {object} newMovie The details about the updated movie
+ * @returns {Promise} Returns the updated movie if updated successfully
+ */
+function updateMovie(id, newMovie) {
+  return new Promise((resolve, reject) => {
```
Comment (svilenvelikov) [inline models/movie.model.js]: Same as here https://github.com/MikeHitrov/MoviesAPI/pull/1/files#r540035867

Comment (svilenvelikov) [reply models/movie.model.js]: See the other methods as well

File: controllers/movie.controller.js [hunk 10]
```diff
@@ -1,25 +1,61 @@
...
+ */
 const updateMovie = (id, parameters) => {
   return movieService.updateMovie(id, parameters);
 };
 
+/**
+ * This function deletes a movie by its id
+ *
+ * @param {number} id The movie id
+ * @returns {undefined}
```
Comment (svilenvelikov) [inline controllers/movie.controller.js]: It's not a standart but it's a practise to return the ID of the deleted entity.

File: routes/movie.routes.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
+const express = require("express");
+const router = express.Router();
+const middlewares = require("../helpers/middlewares");
+const controller = require("../controllers/movie.controller");
+
+// Routes
+module.exports = router;
+
+router.get("/", async (req, res) => {
```
Comment (svilenvelikov) [inline routes/movie.routes.js]: Promise and async are still mixed. Choose one and stick to it!

File: services/user.service.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+/**
+ * This function returns all the movies
+ *
+ * @returns {Array} Returns all the movies
+ */
+function getMovies() {
+  return new Promise((resolve, reject) => {
+    if (movies.length === 0) {
+      reject({
```
Comment (svilenvelikov) [inline services/user.service.js]: Rejection usually means an error of some kind. Here though lack of movies in the store is not an error, so just resolve with empty array and leave the client to decide how to handle it.

File: services/user.service.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+    id = { id: movie.id };
+    const date = {
+      createdAt: movie.createdAt,
+      updatedAt: newDate(),
+    };
+
+    movies[index] = { ...id, ...date, ...newMovie };
+    writeJSONFile(movies);
+    resolve(movies[id]);
```
Comment (svilenvelikov) [inline services/user.service.js:69]: I think that this is wrong. You keep movies in a json array and should be accessing them by index and not by movie ID, right?

File: routes/movie.routes.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  controller
+    .getAllMovies()
+    .then((movies) => res.json(movies))
+    .catch((err) => res.json(err));
+});
+
+router.get("/:id", middlewares.mustBeInteger, async (req, res) => {
```
Comment (svilenvelikov) [inline routes/movie.routes.js]: I like how you integrated a declarative parameter validation using express middlware. Although IDs in most cases are represented as strings usually UUIDs and not an integer, and the reason is that the integer type would introduce a size limit which is impractical.

File: services/user.service.js [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+/**
+ * This function deletes a movie by given id and rewrites the JSON file
+ *
+ * @param {number} id The movie id
+ * @returns {undefined}
+ */
+function deleteMovie(id) {
+  return helper
+    .searchByIdInArray(movies, Number(id))
```
Comment (svilenvelikov) [inline services/user.service.js]: Isn't the ID guaranteed to be a number here?!

Comment (svilenvelikov) [inline services/user.service.js]: I think you don't need this `searchByIdInArray` call here as you don't use the result. You just need to filter out the deleted movie, update the store and return the id of the deleted movie.

File: services/user.service.js [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    .then((index) => {
+      movies = movies.filter((m) => m.id != id);
+      writeJSONFile(movies);
+      resolve(undefined);
+    })
+    .catch((err) => {
```
Comment (svilenvelikov) [inline services/user.service.js]: This catch is redundant here IMO. You are handling the errors in the controller anyway.

File: helpers/helper.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+ * @param  {string} genre The genre that we want to filter
+ * @returns {Array} The movies that are filtered based on the genre
+ */
+function filterByGenre(array, genre) {
+  return new Promise((resolve, reject) => {
+    let rows = array.find((m) => m.genre == genre);
+
+    if (!rows) {
+      reject({
+        message: "There are not any films with this genre.",
```
Comment (svilenvelikov) [inline helpers/helper.js]: `movies` instead of `films`! Try to be consistent!

File: services/user.service.js [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+/**
+ * This function retuns movies filtered by given genre
+ *
+ * @param {string} genre The wanted genre
+ * @returns {Promise} Returns all the filtered movies
+ */
+function getMoviesByGenre(genre) {
+  return new Promise((resolve, reject) => {
```
Comment (svilenvelikov) [inline services/user.service.js]: Redundant wrapping in a promise. The `helper.filterByGenre` already returns a promise.

File: services/user.service.js [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+    helper
+      .filterByGenre(movies, genre)
+      .then((movies) => resolve(movies))
```
Comment (svilenvelikov) [inline services/user.service.js]: Redudndant `then` and `catch`. You don't do anything with the filtered movies, so just return the result from `helper.filterByGenre`

File: services/user.service.js (hunk 13 above)
Comment (svilenvelikov) [reply services/user.service.js:69]: ahh I just saw that you are actually generating IDs based on the actual index, so this seems to be correct. Still kind of awkward thou

File: tests/movie.model.test.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
+const service = require("../services/user.service");
```
Comment (svilenvelikov) [inline tests/movie.model.test.js]: It'll be good if you include tests for the corner cases as well and not just the stright scenarios. What is happeining when you try to delete a movie which does not exists, ot the store is empty, when you search for movie which does not exists, etc.
You can also stub the service `writeJSONFile` method to throw error and test that it's properly handled.

File: README.md [hunk 21]
```diff
@@ -0,0 +1,12 @@
+MoviesAPI
+A simple API, which provides CRUD operations over movies model.
+
+# Before running the app:
+
+`npm install`<br/>
+`yarn install` - jest is unavailable for npm (throws an error)
```
Comment (svilenvelikov) [inline README.md]: I'm qurious about this. According to their ddocumentation they have their release in npm registry and we use that one too.

**Review by svilenvelikov — APPROVED** [review]

File: README.md (hunk 21 above)
Comment (MikeHitrov) [reply README.md]: That's the error I get when try to install jest using the command - npm i jest:
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm ERR! Maximum call stack size exceeded

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/mihailhitrov/.npm/_logs/2020-12-21T14_33_43_119Z-debug.log

Comment (MikeHitrov) [reply README.md]: I edited the README file to use yarn install for installing depenencies.

Comment (svilenvelikov) [reply README.md]: Here is what I get
```
/testjest$ npm i jest
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^2.1.2 (node_modules/jest-haste-map/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.2.1: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN testjest@1.0.0 No description
npm WARN testjest@1.0.0 No repository field.

+ jest@26.6.3
added 511 packages from 350 contributors and audited 512 packages in 10.861s

23 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```
